dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/OptionSetObjectBundleHook.java

```diff
@@ -0,0 +1,66 @@
+package org.hisp.dhis.dxf2.metadata.objectbundle.hooks;
+
+/*
+ * Copyright (c) 2004-2020, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import org.hibernate.SessionFactory;
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.dxf2.metadata.objectbundle.ObjectBundle;
+import org.hisp.dhis.option.OptionSet;
+import org.springframework.stereotype.Component;
+
+@Component
+public class OptionSetObjectBundleHook
+    extends AbstractObjectBundleHook
+{
+    private final SessionFactory sessionFactory;
+
+    public OptionSetObjectBundleHook( SessionFactory sessionFactory )
+    {
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Override
+    public <T extends IdentifiableObject> void postCreate( T persistedObject, ObjectBundle bundle )
+    {
+        if ( !( persistedObject instanceof OptionSet ) )
+        {
+            return;
+        }
+
+        OptionSet optionSet = (OptionSet) persistedObject;
+
+        optionSet.getOptions().forEach( option -> {
```

Comment:
I am worried this can be quite slow, would be good to test on ICD10 option set with 14 000 codes. Will that mean 1400 writes?

Reply:
yes agree, let me look for a workaround solution for this.

Reply:
So we doesn't  need update every of those those options with optionSet = null, instead just refresh the OptionSet entity and hibernate will collect and refresh the options itself. Those are just select queries by primary keys. 

`* INFO  2020-02-03T14:24:25,004 [Level: INFO, category: METADATA_IMPORT, time: Mon Feb 03 14:24:25 ICT 2020, message: (system) Creating 1 object(s) of type OptionSet] (InMemoryNotifier.java [taskScheduler-13])`

`Hibernate: 
    select
        nextval ('hibernate_sequence')
Hibernate: 
    insert 
    into
        optionset
        (uid, code, created, lastUpdated, lastupdatedby, name, translations, valueType, version, attributeValues, userid, publicAccess, optionsetid) 
    values
        (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`

`Hibernate: 
    update
        optionvalue 
    set
        optionsetid=?,
        sort_order=? 
    where
        optionvalueid=?`
`Hibernate: 
    update
        optionvalue 
    set
        optionsetid=?,
        sort_order=? 
    where
        optionvalueid=?`

`Hibernate: 
    select
        option0_.optionvalueid as optionva1_218_0_,
        option0_.uid as uid2_218_0_,
        option0_.code as code3_218_0_,
        option0_.name as name4_218_0_,
        option0_.created as created5_218_0_,
        option0_.lastUpdated as lastUpda6_218_0_,
        option0_.sort_order as sort_ord7_218_0_,
        option0_.description as descript8_218_0_,
        option0_.formName as formName9_218_0_,
        option0_.style as style10_218_0_,
        option0_.translations as transla11_218_0_,
        option0_.optionsetid as options12_218_0_,
        option0_.attributeValues as attribu13_218_0_ 
    from
        optionvalue option0_ 
    where
        option0_.optionvalueid=?`

`Hibernate: 
    select
        option0_.optionvalueid as optionva1_218_0_,
        option0_.uid as uid2_218_0_,
        option0_.code as code3_218_0_,
        option0_.name as name4_218_0_,
        option0_.created as created5_218_0_,
        option0_.lastUpdated as lastUpda6_218_0_,
        option0_.sort_order as sort_ord7_218_0_,
        option0_.description as descript8_218_0_,
        option0_.formName as formName9_218_0_,
        option0_.style as style10_218_0_,
        option0_.translations as transla11_218_0_,
        option0_.optionsetid as options12_218_0_,
        option0_.attributeValues as attribu13_218_0_ 
    from
        optionvalue option0_ 
    where
        option0_.optionvalueid=?`

`Hibernate: 
    select
        optionset0_.optionsetid as optionse1_215_1_,
        optionset0_.uid as uid2_215_1_,
        optionset0_.code as code3_215_1_,
        optionset0_.created as created4_215_1_,
        optionset0_.lastUpdated as lastUpda5_215_1_,
        optionset0_.lastupdatedby as lastupda6_215_1_,
        optionset0_.name as name7_215_1_,
        optionset0_.translations as translat8_215_1_,
        optionset0_.valueType as valueTyp9_215_1_,
        optionset0_.version as version10_215_1_,
        optionset0_.attributeValues as attribu11_215_1_,
        optionset0_.userid as userid12_215_1_,
        optionset0_.publicAccess as publicA13_215_1_,
        options1_.optionsetid as options12_218_3_,
        options1_.optionvalueid as optionva1_218_3_,
        options1_.sort_order as sort_ord7_3_,
        options1_.optionvalueid as optionva1_218_0_,
        options1_.uid as uid2_218_0_,
        options1_.code as code3_218_0_,
        options1_.name as name4_218_0_,
        options1_.created as created5_218_0_,
        options1_.lastUpdated as lastUpda6_218_0_,
        options1_.sort_order as sort_ord7_218_0_,
        options1_.description as descript8_218_0_,
        options1_.formName as formName9_218_0_,
        options1_.style as style10_218_0_,
        options1_.translations as transla11_218_0_,
        options1_.optionsetid as options12_218_0_,
        options1_.attributeValues as attribu13_218_0_ 
    from
        optionset optionset0_ 
    left outer join
        optionvalue options1_ 
            on optionset0_.optionsetid=options1_.optionsetid 
    where
        optionset0_.optionsetid=?`